File: wiki-java/src/main/java/io/github/freya022/wiki/java/commands/slash/SlashReport.java
package io.github.freya022.wiki.java.commands.slash;

import io.github.freya022.botcommands.api.commands.annotations.Command;
import io.github.freya022.botcommands.api.commands.application.slash.GuildSlashEvent;
import io.github.freya022.botcommands.api.commands.application.slash.annotations.JDASlashCommand;
import io.github.freya022.botcommands.api.commands.application.slash.annotations.SlashOption;
import io.github.freya022.botcommands.api.modals.ModalEvent;
import io.github.freya022.botcommands.api.modals.Modals;
import io.github.freya022.botcommands.api.modals.annotations.ModalData;
import io.github.freya022.botcommands.api.modals.annotations.ModalHandler;
import io.github.freya022.botcommands.api.modals.annotations.ModalInput;
import net.dv8tion.jda.api.components.label.Label;
import net.dv8tion.jda.api.components.textinput.TextInput;
import net.dv8tion.jda.api.components.textinput.TextInputStyle;
import net.dv8tion.jda.api.entities.Member;

// --8<-- [start:report-java]
@Command
public class SlashReport {
    private static final String MODAL_NAME = "SlashReport: modal";
    private static final String INPUT_MESSAGE = "SlashReport: modal message";

    private final Modals modals;

    public SlashReport(Modals modals) {
        this.modals = modals;
    }

    @JDASlashCommand(name = "report", description = "Reports a member")
    public void onSlashReport(
            GuildSlashEvent event,
            @SlashOption(description = "The member to report") Member member
    ) {
        var modal = modals.create("Report a member")
                .addComponents(
                        Label.of("Report message",
                                TextInput.create(INPUT_MESSAGE, TextInputStyle.PARAGRAPH)
                                        .setMinLength(10)
                                        .build()
                        )
                )
                .bindTo(MODAL_NAME, member)
                .build();

        event.replyModal(modal).queue();
    }

    @ModalHandler(MODAL_NAME)
    public void onReportModal(
            ModalEvent event,
            @ModalData Member member,
            @ModalInput(INPUT_MESSAGE) String message
    ) {
        event.replyFormat("""
                                You reported %s with the following message:
                                %s
                                """,
                        member.getAsMention(),
                        message)
                .setEphemeral(true)
                .queue();
    }
}
// --8<-- [end:report-java]

File: wiki-kotlin/src/main/kotlin/io/github/freya022/wiki/commands/slash/SlashReport.kt
package io.github.freya022.wiki.commands.slash

import dev.freya02.botcommands.jda.ktx.components.TextInput
import io.github.freya022.botcommands.api.commands.annotations.Command
import io.github.freya022.botcommands.api.commands.application.slash.GuildSlashEvent
import io.github.freya022.botcommands.api.commands.application.slash.annotations.JDASlashCommand
import io.github.freya022.botcommands.api.commands.application.slash.annotations.SlashOption
import io.github.freya022.botcommands.api.modals.ModalEvent
import io.github.freya022.botcommands.api.modals.Modals
import io.github.freya022.botcommands.api.modals.annotations.ModalData
import io.github.freya022.botcommands.api.modals.annotations.ModalHandler
import io.github.freya022.botcommands.api.modals.annotations.ModalInput
import io.github.freya022.botcommands.api.modals.create
import net.dv8tion.jda.api.components.textinput.TextInputStyle
import net.dv8tion.jda.api.entities.Member

// --8<-- [start:report-kotlin]
@Command
class SlashReport(private val modals: Modals) {
    @JDASlashCommand(name = "report", description = "Reports a member")
    fun onSlashReport(
        event: GuildSlashEvent,
        @SlashOption(description = "The member to report") member: Member,
    ) {
        val modal = modals.create("Report a member") {
            label("Report message") {
                child = TextInput(INPUT_MESSAGE, TextInputStyle.PARAGRAPH) {
                    minLength = 10
                }
            }

            bindTo(MODAL_NAME, member)
        }

        event.replyModal(modal).queue()
    }

    @ModalHandler(MODAL_NAME)
    fun onReportModal(
        event: ModalEvent,
        @ModalData member: Member,
        @ModalInput(INPUT_MESSAGE) message: String,
    ) {
        event.reply("""
            You reported ${member.asMention} with the following message:
            $message
        """.trimIndent())
            .setEphemeral(true)
            .queue()
    }

    companion object {
        private const val MODAL_NAME = "SlashReport: modal"
        private const val INPUT_MESSAGE = "SlashReport: modal message"
    }
}
// --8<-- [end:report-kotlin]
